### PR TITLE
Hackathon/edge mesh

### PR DESCRIPTION
**Pull request by Zhouyuan-Chen**
Just add dummy .hpp/.cpp and modified cmake files.

### CODE REVIEW

**Comment by Zhouyuan-Chen** [comment]
fixed problems

File: src/wmtk/EdgeMesh.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
+#include "EdgeMesh.hpp"
+namespace wmtk {
+Tuple EdgeMesh::vertex_tuple_from_id(long id) const
+{
+    throw("this function has not been implemented! -- EdgeMesh.hpp/EdgeMesh.cpp");
+    return Tuple(-1, -1, -1, id, get_cell_hash_slow(id));
+}
+
+EdgeMesh::EdgeMesh()
+    : Mesh(0)
```
Comment (mtao) [inline src/wmtk/EdgeMesh.cpp:10]: this should be Mesh(1) because the highest simplex is an edge (a 1-simplex)

File: src/wmtk/EdgeMesh.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+{
+    throw("this function has not been implemented! -- EdgeMesh.hpp/EdgeMesh.cpp");
+    // every point is on the interior as it has no boundary simplices
+    return false;
+}
+
+void EdgeMesh::initialize(long count)
+{
+    throw("this function has not been implemented! -- EdgeMesh.hpp/EdgeMesh.cpp");
+    set_capacities({count});
```
Comment (mtao) [inline src/wmtk/EdgeMesh.cpp:49]: ah i suspect you borrowed your initialization from PointMesh - for EdgeMesh initialization needs to be a collection of pairs of indices (the endpoints of each edge). The capacities should be set as `{number of vertices, number of edges}`

**Review by mtao — COMMENTED** [review]
this should probably be a PR draft. I started leaving some comments and then realized this is mostly a copy-paste of PointMesh :)

File: src/wmtk/Types.hpp [hunk 3]
```diff
@@ -4,13 +4,14 @@
 
 namespace wmtk {
 using RowVectors3l = Eigen::Matrix<long, Eigen::Dynamic, 3>;
+using RowVectors2l = Eigen::Matrix<long, Eigen::Dynamic, 2>;
```
Comment (daniel-zint) [inline src/wmtk/Types.hpp:7]: Maybe sort the `using`s. 2 first, 3 after, and so on.

**Review by daniel-zint — COMMENTED** [review]
Looks all good so far. I did not go through the functions labeled as not tested yet. I guess the next step is to add tests and make sure all these functions do what they are supposed to do.

**Comment by Zhouyuan-Chen** [comment]
I uploaded the current code and thought the main skeleton was done, but still, it should have some problems that need to be fixed.
Oh, the files related to tests are not finished, so compiled could be passed.
I wrote comments for some specific implementations that I am not sure about.

**Comment by Zhouyuan-Chen** [comment]
For now, all functions under EdgeMesh have a "throw" at the beginning of the function. Please remove it before using it. Therefore, the first simple tests fail.

File: src/wmtk/utils/edgemesh_topology_initialization.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        if (complete_VE[i].size() == 1) {
+            // boundary vertex
+            if (E(complete_VE[i][0], 0) == i) {
+                EE(complete_VE[i][0], 0) = -1;
+            } else {
+                EE(complete_VE[i][0], 1) = -1;
+            }
+        } else {
+            // non-boundary vertex
+            for (int k = 0; k < 2; ++k) {
```
Comment (daniel-zint) [inline src/wmtk/utils/edgemesh_topology_initialization.cpp:41]: `long` (just for consistency)

File: src/wmtk/TriMesh.cpp [hunk 5]
```diff
@@ -337,11 +337,12 @@ bool TriMesh::is_connectivity_valid() const
...
         int cnt = 0;
         for (long j = 0; j < 3; ++j) {
-            if (fe_accessor.index_access().vector_attribute(ef_accessor.index_access().scalar_attribute(i))[j] == i) {
+            if (fe_accessor.index_access().vector_attribute(
+                    ef_accessor.index_access().scalar_attribute(i))[j] == i) {
                 cnt++;
             }
         }
-        if (cnt != 1) {
+        if (cnt == 0) {
```
Comment (daniel-zint) [inline src/wmtk/TriMesh.cpp:345]: I assume these changes are not related to this PR. Did you maybe pull another branch into this one? It might make sense to pull the main branch into this one once more to clean up changes that are actually from some other branch.

File: tests/test_1d_operations.cpp [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+{
+    SECTION("single line")
+    {
+        DEBUG_EdgeMesh m = single_line();
+        REQUIRE(m.is_connectivity_valid());
+
+        const long edge_id = 0;
+        Tuple edge = m.tuple_from_edge_id(edge_id);
+        REQUIRE(m.is_valid_slow(edge));
+        Accessor<long> hash_accessor = m.get_cell_hash_accessor();
```
Comment (daniel-zint) [inline tests/test_1d_operations.cpp:33]: move the `hash_accessor` one line up, then you can use `is_valid`

File: src/wmtk/EdgeMeshOperationExecutor.hpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    Accessor<long> ev_accessor;
+    Accessor<long> ve_accessor;
+    Accessor<long>& hash_accessor;
+
+    /**
+     * @brief gather all simplices that are deleted in a split
+     *
+     * The deleted simplices are exactly the open star of the edge
+     */
+    static const std::array<std::vector<long>, 2> get_split_simplices_to_delete(
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.hpp:25]: Please be more specific in the documentation what is returned (I know the other operation executors are just as badly documented, but please, be better 🙂)

File: src/wmtk/EdgeMeshOperationExecutor.hpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        const EdgeMesh& m);
+
+    /**
+     * @brief gather all simplices that are deleted in a collapse
+     *
+     * The deleted simplices are the intersection of the open star of the vertex and the closed star
+     * of the edge. This comes down to one vertex, three edges, and two faces if the edge is on the
+     * interior. On the boundary it is one vertex, two edges, and one face.
+     */
+    static const std::array<std::vector<long>, 2> get_collapse_simplices_to_delete(
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.hpp:36]: The documentation here does not make sense for an edge mesh.

File: src/wmtk/EdgeMeshOperationExecutor.hpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        const Tuple& tuple,
+        const EdgeMesh& m);
+
+    const std::array<long, 2>& incident_vids() const { return m_spine_vids; }
+
+    long operating_edge_id() const { return m_operating_edge_id; }
+
+    Tuple split_edge();
+    Tuple collapse_edge();
+    Tuple split_edge_single_mesh();
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.hpp:46]: I feel like these `_single_` functions should be private. Correct me if I am wrong.

File: src/wmtk/EdgeMeshOperationExecutor.cpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+
+void EdgeMesh::EdgeMeshOperationExecutor::update_hash_in_map(EdgeMesh& child_mesh)
+{
+    // TODO: Implement for multi_mesh in the future
+}
+
+Tuple EdgeMesh::EdgeMeshOperationExecutor::collapse_edge()
+{
+    return collapse_edge_single_mesh();
+    // TODO: Implement for multi_mesh in the future
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.cpp:170]: If the mesh is part of a multi-mesh, throw an exception.

File: src/wmtk/EdgeMeshOperationExecutor.cpp [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+    assert(new_eids.size() == 2);
+
+    const long local_vid = m_mesh.is_ccw(m_operating_tuple) ? 0 : 1;
+
+    // update ee
+    {
+        // for 2 new edges
+        auto ee_new_0 = ee_accessor.index_access().vector_attribute(new_eids[0]);
+        auto ee_new_1 = ee_accessor.index_access().vector_attribute(new_eids[1]);
+        ee_new_0[(local_vid + 1) % 2] = new_eids[1];
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.cpp:109]: `(local_vid + 1) % 2` can be written as `local_vid ^ 1`

File: src/wmtk/EdgeMeshOperationExecutor.cpp [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    ids[0].emplace_back(m.id_vertex(tuple));
+    ids[1].emplace_back(m.id_edge(tuple));
+    return ids;
+}
+
+std::vector<Tuple> EdgeMesh::EdgeMeshOperationExecutor::prepare_operating_tuples_for_child_meshes()
+    const
+{
+    // this function is designed as a helper for multi_mesh
+    return MultiMeshManager::map_edge_tuple_to_all_children(m_mesh, m_operating_tuple);
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.cpp:80]: should throw an exception if not tested

File: src/wmtk/EdgeMesh.hpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    Tuple collapse_edge(const Tuple& t, Accessor<long>& hash_accessor) override;
+
+    Tuple switch_tuple(const Tuple& tuple, PrimitiveType type) const override;
+
+    bool is_ccw(const Tuple& tuple) const override;
+    bool is_boundary(const Tuple& tuple) const override;
+    bool is_boundary_vertex(const Tuple& tuple) const override;
+    bool is_boundary_edge(const Tuple& tuple) const override
+    {
+        throw("This function doesn't make sense for edgemesh");
```
Comment (daniel-zint) [inline src/wmtk/EdgeMesh.hpp:32]: `EdgeMesh`

File: src/wmtk/EdgeMesh.cpp [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+        }
+    }
+
+    // EV and EE
+    for (long i = 0; i < capacity(PrimitiveType::Edge); ++i) {
+        if (e_flag_accessor.index_access().scalar_attribute(i) == 0) {
+            wmtk::logger().debug("Edge {} is deleted", i);
+            continue;
+        }
+        // TODO: need to handle cornor case (self-loop)
```
Comment (daniel-zint) [inline src/wmtk/EdgeMesh.cpp:249]: You have a test case for that, right? Then this test case should not pass. You can flag it with `[.]` but it should contain a `REQUIRE(false);`.

**Review by daniel-zint — CHANGES_REQUESTED** [review]
This is a massive PR. My main concern is that there is a lot of code of which I am not sure if it was tested. I added a few comments regarding that. Especially, if you copy code, throw an exception in every function. The fact that they look like they do something reasonable but might have not been tested is extremely dangerous. So please make sure that **every** function is at least called once in a test or throws an exception.

File: src/wmtk/EdgeMeshOperationExecutor.cpp (hunk 10 above)
Comment (JcDai) [reply src/wmtk/EdgeMeshOperationExecutor.cpp:170]: This function is already used in tests.

File: tests/tools/DEBUG_EdgeMesh.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+//     : EdgeMesh(m)
+// {}
+DEBUG_EdgeMesh::DEBUG_EdgeMesh(EdgeMesh&& m)
+    : EdgeMesh(std::move(m))
+{}
+
+
+bool DEBUG_EdgeMesh::operator==(const DEBUG_EdgeMesh& o) const
+{
+    return static_cast<const DEBUG_EdgeMesh&>(*this) == static_cast<const DEBUG_EdgeMesh&>(o);
```
[the remark this answers is not in the document]
Comment (JcDai) [reply tests/tools/DEBUG_EdgeMesh.cpp:16]: Not sure, in DEBUG_TriMesh it casts to a DEBUG_TriMesh

File: tests/test_topology.cpp [hunk 16]
```diff
@@ -334,3 +335,158 @@ TEST_CASE("topology_of_tet_bunny", "[topology][3D]")
...
+
+    // 2. Test relationship between EV and EE
+    for (int i = 0; i < EE.rows(); ++i) {
+        for (int j = 0; j < 2; ++j) {
+            long nb = EE(i, j);
+            if (nb < 0) continue;
+
+            CHECK((EE.row(nb).array() == i).any());
+
+            // TODO add checks
```
[the remark this answers is not in the document]
Comment (JcDai) [reply tests/test_topology.cpp:365]: ok I've opened the issue

File: src/wmtk/TriMesh.cpp (hunk 5 above)
Comment (JcDai) [reply src/wmtk/TriMesh.cpp:329]: I created an issue for this change. It is actually not determined yet (check for glued edges in trimesh). 
I redo the code in this PR

Comment (JcDai) [reply src/wmtk/TriMesh.cpp]: I cannot redo this change. Many tests in 2d need this change. It is not from another branch. We did this change when we realized the corner case in 2d is not handled as it was in 1d so we did this change here.

File: src/wmtk/EdgeMesh.cpp [hunk 17]
```diff
@@ -78,6 +78,12 @@ Tuple EdgeMesh::switch_tuple(const Tuple& tuple, PrimitiveType type) const
         auto ee = ee_accessor.vector_attribute(tuple);
 
         long gcid_new = ee(tuple.m_local_vid);
+
+        // TODO: This is for special case self-loop, just to make sure the local vid of the returned
+        // tuple is the same as the input. (When doing double-switch this is needed)
+        if (gcid_new == tuple.m_global_cid) return tuple;
```
Comment (daniel-zint) [inline src/wmtk/EdgeMesh.cpp:84]: We don't want one line if statements. Put brackets around the return statement.

File: src/wmtk/EdgeMeshOperationExecutor.hpp [hunk 18]
```diff
@@ -20,7 +20,7 @@ class EdgeMesh::EdgeMeshOperationExecutor
     /**
      * @brief gather all simplices that are deleted in a split
      *
-     * The deleted simplices are exactly the open star of the edge
+     * The deleted simplex is the edge itself
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.hpp:23]: That helps but you should also state the structure of `std::array<std::vector<long>, 2>`. What is contained in that array and in the vectors.

File: tests/test_tuple_1d.cpp [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+        CHECK(m.id(vertices[4], PrimitiveType::Vertex) == 4);
+        CHECK(m.id(vertices[5], PrimitiveType::Vertex) == 5);
+        CHECK(m.id(vertices[0], PrimitiveType::Edge) == 0);
+        CHECK(m.id(vertices[5], PrimitiveType::Edge) == 4);
+    }
+    SECTION("edges")
+    {
+        const std::vector<Tuple> edges = m.get_all(PrimitiveType::Edge);
+        REQUIRE(edges.size() == 5);
+        CHECK(m.id(edges[0], PrimitiveType::Edge) == 0);
```
Comment (daniel-zint) [inline tests/test_tuple_1d.cpp:125]: You can write those `CHECK`s as for-loop.

File: tests/test_tuple_1d.cpp [hunk 20]
```diff
@@ -0,0 +1,319 @@
...
+                switch (rand() % 2) {
+                case 0: t = m.switch_tuple(t, PrimitiveType::Vertex); break;
+                case 1:
+                    if (!m.is_boundary(t)) {
+                        t = m.switch_tuple(t, PrimitiveType::Edge);
+                    }
+                    break;
+                default: break;
+                }
+                CHECK(m.is_valid_slow(t));
```
Comment (daniel-zint) [inline tests/test_tuple_1d.cpp:198]: Also here it would be nice if you could just get the hash accessor once and reuse it. You can even define it outside of the sections.

File: tests/test_tuple_1d.cpp [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+    // edges
+    const std::vector<Tuple> edges = m.get_all(PrimitiveType::Edge);
+    for (const auto& t : edges) {
+        const Tuple t_after_v = m.switch_vertex(m.switch_vertex(t));
+        CHECK(t == t_after_v);
+        if (!m.is_boundary(t)) {
+            const Tuple t_after_e = m.switch_edge(m.switch_edge(t));
+            CHECK(t == t_after_e);
+        }
+    }
```
Comment (daniel-zint) [inline tests/test_tuple_1d.cpp:318]: The tests look very clean now. Nice!

File: tests/tools/DEBUG_EdgeMesh.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (daniel-zint) [reply tests/tools/DEBUG_EdgeMesh.cpp:16]: https://github.com/wildmeshing/wildmeshing-toolkit/blob/165270421bd1aa15a2a3ffa5de43e4ddcf21b1c5/tests/tools/DEBUG_TriMesh.cpp#L16

No it doesn't. That would also not make any sense because the cast wouldn't do anything. 

Actually, you created an endless loop here. That also means that this function was never used and never tested. Therefore, throw an exception.

**Review by daniel-zint — COMMENTED** [review]
Looks very good already. I am just still a bit concerned that there are functions that were implemented but never tested.

`DEBUG_EdgeMesh::operator==` needs to be fixed. There are a few more minor comments. I think the PR is almost there.

File: src/wmtk/EdgeMesh.cpp [hunk 22]
```diff
@@ -81,8 +81,9 @@ Tuple EdgeMesh::switch_tuple(const Tuple& tuple, PrimitiveType type) const
 
         // TODO: This is for special case self-loop, just to make sure the local vid of the returned
```
Comment (daniel-zint) [inline src/wmtk/EdgeMesh.cpp:82]: Is this TODO still a TODO?

**Review by daniel-zint — APPROVED** [review]
lgtm

**Comment by mtao** [comment]
What is left before we can merge this? I'll trust daniel's review on this, but what's left to finish off his comments?
If you're working on things that are not part of his comments this PR those changes will need to be deleted.

**Comment by JcDai** [comment]
> What is left before we can merge this? I'll trust daniel's review on this, but what's left to finish off his comments? If you're working on things that are not part of his comments this PR those changes will need to be deleted.

I think daniel's comments are finished. There was one merge main branch to our branch after finishing the comments. I think nothing else there?

File: src/wmtk/EdgeMeshOperationExecutor.cpp [hunk 23]
```diff
@@ -78,7 +78,9 @@ std::vector<Tuple> EdgeMesh::EdgeMeshOperationExecutor::prepare_operating_tuples
 {
     // this function is designed as a helper for multi_mesh
     throw("this function is not tested");
-    return MultiMeshManager::map_edge_tuple_to_all_children(m_mesh, m_operating_tuple);
+    return m_mesh.m_multi_mesh_manager.map_edge_tuple_to_all_children(
```
Comment (daniel-zint) [inline src/wmtk/EdgeMeshOperationExecutor.cpp:81]: This `return` causes the error in the Windows Debug compilation. Just comment it out.

**Review by daniel-zint — COMMENTED** [review]
Just fix that compilation issue and then let's finish this PR.
